src/crimson/os/seastore/onode_manager/simple-fltree/onode_node.h

```diff
@@ -0,0 +1,962 @@
+                                    onode_t,
+                                    child_addr_t>;
+  using const_item_t = std::conditional_t<NodeType == ntype_t::leaf,
+                                          const onode_t&,
+                                          child_addr_t>;
+  // XXX: magic numbers here
+  static constexpr bool item_in_key = (N == 3 &&
+                                       NodeType == ntype_t::inner);
+  using partial_key_t = std::conditional_t<N < 3,
+                                           variable_sized_key_t,
+                                           empty_key_t>;
+
+  std::pair<const key_t&, const partial_key_t&>
+  key_at(unsigned slot) const;
+
+  // update an existing oid with the specified item
+  ghobject_t update_oid_with_slot(unsigned slot, const ghobject_t& oid) const;
+  const_item_t item_at(const key_t& key) const;
+  void dump(std::ostream& os) const;
+
+  // for debugging only.
+  static constexpr bool is_leaf() {
+    return node_type == ntype_t::leaf;
+  }
+
+  bool _is_leaf() const {
+    return tag.is_leaf();
+  }
+
+  char* from_end(uint16_t offset);
+  const char* from_end(uint16_t offset) const;
+  uint16_t used_space() const;
+  uint16_t free_space() const {
+    return capacity() - used_space();
+  }
+  static uint16_t capacity();
+  // TODO: if it's allowed to update 2 siblings at the same time, we can have
+  //       B* tree
+  static constexpr uint16_t min_size();
+
+  // @return <minimum bytes to grab, maximum bytes to grab>
+  static constexpr std::pair<int16_t, int16_t> bytes_to_grab(uint16_t size);
+
+  // @return <minimum bytes to push, maximum bytes to push>
+  static constexpr std::pair<int16_t, int16_t> bytes_to_push(uint16_t size);
+
+  size_state_t size_state(uint16_t size) const;
+  bool is_underflow(uint16_t size) const;
+  int16_t size_with_key(unsigned slot, const ghobject_t& oid) const;
+  ordering_t compare_with_slot(unsigned slot, const ghobject_t& oid) const;
+  /// return the slot number of the first slot that is greater or equal to
+  /// key
+  std::pair<unsigned, bool> lower_bound(const ghobject_t& oid) const;
+  static uint16_t size_of_item(const ghobject_t& oid, const item_t& item);
+  bool is_overflow(const ghobject_t& oid, const item_t& item) const;
+  bool is_overflow(const ghobject_t& oid, const OnodeRef& item) const;
+
+  // inserts an item into the given slot, pushing all subsequent keys forward
+  // @note if the item is not embedded in key, shift the right half as well
+  void insert_at(unsigned slot, const ghobject_t& oid, const item_t& item);
+  // used by InnerNode for updating the keys indexing its children when their lower boundaries
+  // is updated
+  void update_key_at(unsigned slot, const ghobject_t& oid);
+
+  // @return the number of element to grab
+  std::pair<unsigned, uint16_t> calc_grab_front(uint16_t min_grab, uint16_t max_grab) const;
+  std::pair<unsigned, uint16_t> calc_grab_back(uint16_t min_grab, uint16_t max_grab) const;
```

Comment:
These two don't have users and I'm not actually sure what their purpose is.  Can you add a comment explaining what they are for and what invariants they maintain?

Reply:
sure. comments added. they will be used by the next PR.